### PR TITLE
JSON validator

### PR DESCRIPTION
**Pull request by anjaf**
This adds the json_schema_validation module with the functionality to validate a USI submission envelope (a JSON file with the entire submission metadata) against the full ArrayExpress submission schema. To perform the validation, the jsonschema python package is used.
For this I have created a summary schema that describes the submission envelope itself and links out to the sub-schemas that describe the individual submittables (study, samples, assays, etc). 
To run the validation only and input any given schema the json_validation.py script can be used. 
Also added a few simple unittests that check if the AE schema can be correctly parsed and the validator runs without errors.

### CODE REVIEW

File: json_validation.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import argparse
+
+from validator.json_schema_validation import validate_submission_json
+from utils.common_utils import file_exists
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('json',
```
Comment (pcm32) [inline json_validation.py:16]: I think convention when using the add_argument bit is to provide options in the form that you're using for the following argument (the schema), so providing a short cut for the argument (-j) and a longer doubled dashed one version of it (--json). I would suggest though to be more explicit at this point on what is the purpose of the json (like json for USI input data), as in the future the script might receive other jsons and then telling them apart will be more painful (where as now you don't have any API commitments to comply as this is the first version).

File: tests/test_json_schema_validation.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        schema_file, schema = load_arrayexpress_submission_schema()
+        # Check the schema content can be parsed
+        self.assertIn("samples", schema["properties"])
+
+
+class TestValidateJson(TestCase):
+
+    def setUp(self):
+        # This test uses a very simple test JSON and test schema
+        wd = os.path.dirname(os.path.realpath(__file__))
```
Comment (pcm32) [inline tests/test_json_schema_validation.py]: if `os.path` is the only thing you'd be using from `os`, you might as well be explicit in the import and then save some characters here.

File: validator/json_schema_validation.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    validator = jsonschema.Draft4Validator(schema, resolver=resolver)
+
+    # Check validity of schema first
+    # We shouldn't need this but for now, running this test before calling the checks on the object
+    schema_errors = validator.check_schema(schema)
+
+    # Validate the submission JSON against the submission schema
+    validation_errors = validator.iter_errors(json_data)
+
+    # Print out all error messages with where and why details
```
Comment (pcm32) [inline validator/json_schema_validation.py:38]: Excellent!

File: validator/json_schema_validation.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    """Find and load the ArrayExpress submission schema as JSON"""
+
+    # Construct path to schema and load as json
+    schema_path = os.path.dirname(json_schemas.__file__)
+    schema_file = os.path.join(schema_path, "arrayexpress_submission_schema.json")
+    schema = read_json_file(schema_file)
+
+    return schema_file, schema
+
+
```
Comment (pcm32) [inline validator/json_schema_validation.py:53]: I don't know out of memory, but I would be surprised if there are no good libraries to deal with tracebacks traversals and error formatting for reporting. Only if you find something adequate of course.

File: README.md [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 
-This repository contains modules for transformation of experimental metadata between [MAGE-TAB](fged.org/projects/mage-tab/) and JSON format used in the [Unified Submissions Interface (USI)](https://github.com/EMBL-EBI-SUBS). 
+This repository contains modules for transformation of experimental metadata between [MAGE-TAB](http://fged.org/projects/mage-tab/) and JSON format used in the [Unified Submissions Interface (USI)](https://github.com/EMBL-EBI-SUBS). 
 
 
 ## Installation and requirements
 
 Python 3.6 interpreter<br>
-Package requirements:git 
+Package requirements:
```
Comment (pcm32) [inline README.md:9]: might be good to have this on a requirements.txt file and make the whole thing pip installable in the near future.

**Review by pcm32 — APPROVED** [review]
Overall very good, some optional comments to consider. Maybe the argument passing is worth improving now that you're on the first versions and there are no current strong dependencies on this.

File: README.md (hunk 5 above)
Comment (anjaf) [reply README.md:9]: Yes, there is a requirements.txt too: https://github.com/ebi-gene-expression-group/usi-arrayexpress/blob/develop/requirements.txt We've made that for the Travis integration. I am keeping the package creation still in the back of my mind :)

File: tests/test_json_schema_validation.py (hunk 2 above)
Comment (anjaf) [reply tests/test_json_schema_validation.py]: OK, done.

File: validator/json_schema_validation.py (hunk 3 above)
Comment (anjaf) [reply validator/json_schema_validation.py:38]: Thanks!

File: validator/json_schema_validation.py (hunk 4 above)
Comment (anjaf) [reply validator/json_schema_validation.py:53]: Ok, will have a look. The traceback here is a list of the elements in the JSON before the element with error(s).

File: json_validation.py (hunk 1 above)
Comment (anjaf) [reply json_validation.py:16]: Yes, being explicit is always good. But I'll leave this for now as it is meant as a simple general-purpose script (with one mandatory input file).

**Pull request merged**